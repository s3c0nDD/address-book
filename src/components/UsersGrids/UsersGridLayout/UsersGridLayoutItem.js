import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';

import Grid from '@material-ui/core/Grid';
import Typography from '@material-ui/core/Typography';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardMedia from '@material-ui/core/CardMedia';
import CardContent from '@material-ui/core/CardContent';
import PersonIcon from '@material-ui/icons/Person';
import EmailIcon from '@material-ui/icons/Email';

const useStyles = makeStyles(theme => ({
  root: {
    maxWidth: 260
  },
  media: {
    height: 80,
    [theme.breakpoints.up('sm')]: {
      height: 100,
    },
    [theme.breakpoints.up('md')]: {
      height: 120,
    },
    [theme.breakpoints.up('lg')]: {
      height: 150,
    }
  }
}));

/**
 * Checks props and decide if should not re-render
 * @param {object} prevProps previous props
 * @param {object} nextProps next props
 * @returns {boolean}
 */
const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.user === nextProps.user;
};

/**
 * User grid layout item component
 * @param {object} props component's props
 * @param {object} props.user user data
 * @param {function} props.onOpenModal user modal open handler
 */
const UsersGridLayoutItem = memo(({ user, onOpenModal }) => {
  const classes = useStyles();

  /**
   * Handle grid layout item media click
   */
  const handleClick = () => {
    onOpenModal(user);
  };

  const fullName = useMemo(() => (
    `${user.name.first} ${user.name.last}`
  ), [user]);

  return (
    <Card className={classes.root}>
      <CardActionArea
        onClick={handleClick}
      >
        <CardMedia
          className={classes.media}
          image={user.picture.large}
          title={`${fullName} face photo`}
        />
        <CardContent>
          <Typography
            gutterBottom
            variant="h5"
            component="h2"
          >
            {fullName}
          </Typography>

          <Grid
            container
            justify="flex-start"
            alignItems="stretch"
          >
            <Grid item xs={3} md={2}>
              <PersonIcon color="primary" />
            </Grid>
            <Grid item xs={9} md={10} zeroMinWidth>
              <Typography
                noWrap
                variant="body2"
                color="textSecondary"
                component="p"
              >
                {user.login.username}
              </Typography>
            </Grid>
          </Grid>

          <Grid container>
            <Grid item xs={3} md={2}>
              <EmailIcon color="primary" />
            </Grid>
            <Grid item xs={9} md={10} zeroMinWidth>
              <Typography
                noWrap
                variant="body2"
                color="textSecondary"
                component="p"
              >
                {user.email}
              </Typography>
            </Grid>
          </Grid>
        </CardContent>
      </CardActionArea>
    </Card>
  );
}, shouldNotRerender);

export default UsersGridLayoutItem;

UsersGridLayoutItem.propTypes = {
  user: PropTypes.shape({
    name: PropTypes.shape({
      first: PropTypes.string.isRequired,
      last: PropTypes.string.isRequired
    }).isRequired,
    picture: PropTypes.shape({
      large: PropTypes.string.isRequired
    }).isRequired,
    login: PropTypes.shape({
      username: PropTypes.string.isRequired
    }).isRequired,
    email: PropTypes.string.isRequired
  }).isRequired,
  onOpenModal: PropTypes.func.isRequired
};