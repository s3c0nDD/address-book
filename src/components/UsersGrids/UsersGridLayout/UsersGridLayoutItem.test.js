import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import CardActionArea from '@material-ui/core/CardActionArea';

import userMockData from '../../mocks/userMock';
import UsersGridLayoutItem from './UsersGridLayoutItem'

describe('UsersGridLayoutItem', () => {
  let shallow;
  let wrapper;
  let openModalMock;
  let userMock = { ...userMockData };

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    openModalMock = jest.fn();
    wrapper = shallow(
      <UsersGridLayoutItem
        user={userMock}
        onOpenModal={openModalMock}
      />
    );
  });

  it('shallowly renders and matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('calls open modal handler on CardActionArea click', () => {
    const actionWrapper = wrapper.find(CardActionArea);
    actionWrapper.props().onClick();
    expect(openModalMock).toBeCalledTimes(1);
    expect(openModalMock).toBeCalledWith(userMock);
  });
});