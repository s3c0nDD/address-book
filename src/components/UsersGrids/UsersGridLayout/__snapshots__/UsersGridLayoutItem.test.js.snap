// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersGridLayoutItem shallowly renders and matches snapshot 1`] = `
<WithStyles(ForwardRef(Card))
  className="makeStyles-root-1"
>
  <WithStyles(ForwardRef(CardActionArea))
    onClick={[Function]}
  >
    <WithStyles(ForwardRef(CardMedia))
      className="makeStyles-media-2"
      image="https://wwww.image.com/data/1"
      title="John Doe face photo"
    />
    <WithStyles(ForwardRef(CardContent))>
      <WithStyles(ForwardRef(Typography))
        component="h2"
        gutterBottom={true}
        variant="h5"
      >
        John Doe
      </WithStyles(ForwardRef(Typography))>
      <WithStyles(ForwardRef(Grid))
        alignItems="stretch"
        container={true}
        justify="flex-start"
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={2}
          xs={3}
        >
          <PersonIcon
            color="primary"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={10}
          xs={9}
          zeroMinWidth={true}
        >
          <WithStyles(ForwardRef(Typography))
            color="textSecondary"
            component="p"
            noWrap={true}
            variant="body2"
          >
            johndoe123
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
      <WithStyles(ForwardRef(Grid))
        container={true}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={2}
          xs={3}
        >
          <EmailIcon
            color="primary"
          />
        </WithStyles(ForwardRef(Grid))>
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={10}
          xs={9}
          zeroMinWidth={true}
        >
          <WithStyles(ForwardRef(Typography))
            color="textSecondary"
            component="p"
            noWrap={true}
            variant="body2"
          >
            johndoe123@google.com
          </WithStyles(ForwardRef(Typography))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </WithStyles(ForwardRef(CardContent))>
  </WithStyles(ForwardRef(CardActionArea))>
</WithStyles(ForwardRef(Card))>
`;
